src/frontend/src/expr/parameter.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::sync::{Arc, RwLock};
+
+use risingwave_common::types::DataType;
+
+use super::Expr;
+
+#[derive(Clone, Debug)]
+pub struct Parameter {
+    pub index: u64,
+    param_types: Arc<RwLock<HashMap<u64, Option<DataType>>>>,
+}
```

Comment:
Since the `Parameter`  always accesses its own index and there are no await points in binding, it seems `Rc<RefCell<Option<DataType>>>` is enough.

Reply:
If we will use Rc<RefCell<Option<DataType>>>, the Parameter will not have Sync tarit so do ExprImpl. It wiil cause the error in other place 🥵:
https://github.com/risingwavelabs/risingwave/blob/53da2e3246adfdc3b59d60186013279858bcb110/src/frontend/src/utils/condition.rs#L91

Reply:
Ok, I got it.